rust/src/countme/repo.rs

```diff
@@ -0,0 +1,131 @@
+fn is_true(string: &str) -> bool {
+    string == "1" || string == "yes" || string == "true" || string == "on"
+}
+
+/// Read all repository configuration files from the default location
+pub fn all() -> Result<Vec<Repo>> {
+    let configs = fs::read_dir(YUM_REPOS_D)
+        .with_context(|| format!("Could not list files in: {}", YUM_REPOS_D))?;
+    let mut repos = Vec::new();
+    for c in configs {
+        let path = match c {
+            Err(e) => {
+                println!("Could not list file in : {}: {}", YUM_REPOS_D, e);
+                continue;
+            }
+            Ok(p) => p.path(),
+        };
+        match from_file(&path) {
+            Err(e) => {
+                println!("Skipped file: {}: {}", path.display(), e)
+            }
+            Ok(mut r) => repos.append(&mut r),
+        }
+    }
+    Ok(repos)
+}
+
+/// Read repository configuration from a file
+fn from_file(path: &PathBuf) -> Result<Vec<Repo>> {
+    let i = Ini::load_from_file(path).with_context(|| "Could not parse as INI".to_string())?;
+    let mut repos = Vec::new();
+    for (sec, prop) in i.iter() {
+        let mut repo = match sec {
+            None => {
+                println!("Skipping config values not attached to a repository");
+                continue;
+            }
+            Some(s) => Repo {
+                name: String::from(s),
+                enabled: false,
+                count_me: false,
+                meta_link: "".to_string(),
+            },
+        };
+        for (k, v) in prop.iter() {
+            match k {
+                "countme" => {
+                    if is_true(v) {
+                        repo.count_me = true
+                    }
+                }
+                "enabled" => {
+                    if is_true(v) {
+                        repo.enabled = true
+                    }
+                }
+                "metalink" => repo.meta_link = String::from(v),
+                _ => {}
+            }
+        }
+        repos.push(repo);
+    }
+    Ok(repos)
+}
+
+impl Repo {
+    /// Returns true if this repo is
+    /// - enabled
+    /// - configured for sending a Count Me request
+    /// - has a non-empty metalink URL
+    pub fn count_me(&self) -> bool {
+        self.enabled && self.count_me && !self.meta_link.is_empty()
+    }
+
+    /// Get the metalink URL for the repo with variables replaced
+    pub fn metalink(&self, version_id: &str) -> String {
+        // Map some Rust architecture constants to specific strings to match dnf
+        // mapping. See:
+        // https://github.com/rpm-software-management/dnf/blob/4.5.2/dnf/rpm/__init__.py#L88
+        let arch = match std::env::consts::ARCH {
+            "arm" => "armhfp",
+            "powerpc" => "ppc",
+            "powerpc64" => "ppc64",
+            "sparc64" => "sparc",
+            "x86" => "i386",
```

Comment:
Let's factor out this mapping into a `utils.rs` helper like `fn get_rpm_basearch()` and also with a TODO that we want to use libdnf for it in the future when we can more properly link the Rust side to the C/C++ side.

Also did you verify `aarch64` and `riscv` match?

Reply:
Yes, `aarch64` and `riscv` match.